source/windows/system_info.c

```diff
@@ -80,58 +86,131 @@ void aws_backtrace_print(FILE *fp, void *call_site_data) {
         goto done;
     }
 
-    SymInitialize_fn *p_SymInitialize = (SymInitialize_fn *)GetProcAddress(dbghelp, "SymInitialize");
-    if (!p_SymInitialize) {
+    s_SymInitialize = (SymInitialize_fn *)GetProcAddress(dbghelp, "SymInitialize");
+    if (!s_SymInitialize) {
         fprintf(stderr, "Failed to load SymInitialize from DbgHelp.dll.\n");
         goto done;
     }
 
-    SymFromAddr_fn *p_SymFromAddr = (SymFromAddr_fn *)GetProcAddress(dbghelp, "SymFromAddr");
-    if (!p_SymFromAddr) {
+    s_SymFromAddr = (SymFromAddr_fn *)GetProcAddress(dbghelp, "SymFromAddr");
+    if (!s_SymFromAddr) {
         fprintf(stderr, "Failed to load SymFromAddr from DbgHelp.dll.\n");
         goto done;
     }
 
-    SymGetLineFromAddr_fn *p_SymGetLineFromAddr =
-        (SymGetLineFromAddr_fn *)GetProcAddress(dbghelp, SymGetLineFromAddrName);
-    if (!p_SymGetLineFromAddr) {
+    s_SymGetLineFromAddr = (SymGetLineFromAddr_fn *)GetProcAddress(dbghelp, SymGetLineFromAddrName);
+    if (!s_SymGetLineFromAddr) {
         fprintf(stderr, "Failed to load " SymGetLineFromAddrName " from DbgHelp.dll.\n");
         goto done;
     }
 
     HANDLE process = GetCurrentProcess();
-    p_SymInitialize(process, NULL, TRUE);
-    void *stack[1024];
-    WORD num_frames = CaptureStackBackTrace(0, 1024, stack, NULL);
+    AWS_FATAL_ASSERT(process);
+    s_SymInitialize(process, NULL, TRUE);
+    return true;
+
+done:
+    if (dbghelp) {
+        FreeLibrary(dbghelp);
+    }
+    return false;
+}
+
+int aws_backtrace(void **frames, size_t size) {
+    return (int)CaptureStackBackTrace(0, (ULONG)size, frames, NULL);
+}
+
+char **aws_backtrace_symbols(void *const *stack, size_t num_frames) {
+    if (!s_init_dbghelp()) {
```

Comment:
Should we do this as a `aws_thread_call_once`?

Reply:
It can fail, so we would crash trying to use the resolved functions. Otherwise, yeah, that'd be the best way.

Reply:
Does that mean that if it does fail, we're going to retry every time one of these functions get called? Could those ever succeed? It might be cleaner to add a have something like this:
```c
if s_try_init_dbghelp() {
    if (one_of_the_functions) {
        return true;
    } else {
        aws_thread_call_once(do_init);
        return one_of_the_functions != NULL;
    }
}
```